### PR TITLE
select_related for MongoEngine

### PR DESCRIPTION
**Pull request by colinhowe**
This adds two methods: select_related and fetch_references. They both do the same thing but one can be used without a queryset (useful if you end up with a list of items).

These methods will go over each document in the list (or queryset) and build up a map of any documents that are referenced (via ReferenceField) and then do as few bulk fetches as possible to get them.

These methods will also work over ListField(ReferenceField) instances.

### CODE REVIEW

File: mongoengine/tests.py [hunk 1]
```diff
@@ -57,3 +57,9 @@ class query_counter(object):
         count = self.db.system.profile.find().count() - self.counter
         self.counter += 1
         return count
+
```
Comment (rozza) [inline mongoengine/tests.py:60]: this needed?

**Comment by rozza** [comment]
I like it - I think it might need some updating to handle dictionaries as well.

https://github.com/colinhowe/mongoengine/blob/2551081ff52b3415ad03cee703c5bc54580aa1ad/mongoengine/base.py#L162
Handles lazily dereferencing of dicts and lists - it feels like these methods should be merged somehow.

File: mongoengine/tests.py (hunk 1 above)
Comment (colinhowe) [inline mongoengine/tests.py:60]: not needed. I added it to make debugging easier and thought it useful enough to keep it. Don't mind if it dies though!

**Comment by colinhowe** [comment]
Yes, handling dictionaries would be good but we don't use that so can't give it as thorough a testing as the rest.

Merging the two different methods would be awesome. I was actually thinking of trying to move some of this into the field classes themselves - I don't like how coupled this is.

**Comment by rozza** [comment]
Updated - dereferencing now works for querysets / documents and when fetching complex fields :)

**Comment by colinhowe** [comment]
You beat me to it! I got distracted by some other stuff at work :(

Looks awesome though!

**Pull request closed without merging**